Step2 : 문자열 덧셈 계산기

codeleesh opened the pull request:
안녕하세요 리뷰어님! 

Step2 문자열 덧셈 계산기 리뷰 요청드립니다.

정답을 찾지 말라고 하셨는데, 자꾸 정답을 찾으려고 하는데 제 자신이 보이네요!

문자열 덧셈을 처리하는 메인 로직에 대해서 꼭 리뷰 받고 싶습니다!

감사합니다.

codeleesh committed:
delete empty.txt

codeleesh committed:
docs: 문자열 덧셈 계산기 기능 구현 목록 정의

codeleesh committed:
feat: 문자열 덧셈 계산기 요구사항 구현
- null 또는 빈 문자열 처리
- 숫자 하나인 경우 처리
- 쉼표구분자인 경우 처리
- 쉼표 또는 콜론 구분자인 경우 처리
- 커스텀 구분자인 경우 처리
- 음수인 경우 Exception 처리

codeleesh committed:
test: 문자열 덧셈 계산기 요구사항의 테스트 구현
- null 또는 빈 문자열 처리
- 숫자 하나인 경우 처리
- 쉼표구분자인 경우 처리
- 쉼표 또는 콜론 구분자인 경우 처리
- 커스텀 구분자인 경우 처리
- 음수인 경우 Exception 처리

codeleesh committed:
remove 사용자 정의 예외 처리 제거

hhlyn0328 commented on README.md:
```diff
@@ -6,4 +6,13 @@
+* [텍스트와 이미지로 살펴보는 온라인 코드 리뷰 과정](https://github.com/next-step/nextstep-docs/tree/master/codereview)
+
+## 2단계 - 문자열 덧셈 계산기
+ - 빈 문자열 또는 null 값을 입력할 경우 0을 반환해야 한다.
```
요구사항 정리를 잘해주셨네요 :+1:

codeleesh replied:
좀 더 세분화하게 적으려고 노력하는데 잘 안됩니다!

hhlyn0328 commented on src/main/java/util/StringAddCalculator.java:
```diff
@@ -0,0 +1,55 @@
+        if(str.contains(COMMA) && str.contains(":") ) return checkCommaAndDelimiter(str);
+        Matcher m = Pattern.compile("//(.)\n(.*)").matcher(str);
+        if(m.find()) return checkCustomDelimiter(m);
+        return result;
```
위의 케이스들을 모두 만족하지 못하면 0이 리턴되는게 맞을까요, 예외가 발생하는게 맞을까요? 

마지막에 기본 구분자로 무조건 처리하도록 하여, 
위의 5개 if 문을 모두 만족하지 않는, 잘못된 형식의 문자열인 경우 예외가 발생하도록 할 것 같아요

codeleesh replied:
네 요구사항의 해당되지 않기 때문에 0이 아니라 예외를 발생시키는 것이 맞을것 같습니다!

hhlyn0328 commented on src/main/java/util/StringAddCalculator.java:
```diff
@@ -0,0 +1,55 @@
+        int result = 0;
+        if( str == null || str.isEmpty() ) return result;
+        if( str.length() == 1 ) return checkOneNumber(str);
+        if( str.contains(COMMA) && !str.contains(":")) return checkComma(str);
```
",|:" 구분자로 콤마 혹은 콜론을 사용할 수 있습니다. 
따라서 콤마는 포함하면서 콜론을 포함하지 않는 경우, 둘 다 포함하는 경우를 나눠서 체크하지 않아도 될 것 같아요!

hhlyn0328 commented on src/main/java/util/StringAddCalculator.java:
```diff
@@ -0,0 +1,55 @@
+
+public class StringAddCalculator {
+
+    public static final String COMMA = ",";
```
콤마외에도 상수로 선언할 수 있는 문자열이나 매직넘버가 더 있을 것 같아요!

hhlyn0328 commented on src/main/java/util/StringAddCalculator.java:
```diff
@@ -0,0 +1,55 @@
+        if( str.length() == 1 ) return checkOneNumber(str);
+        if( str.contains(COMMA) && !str.contains(":")) return checkComma(str);
+        if(str.contains(COMMA) && str.contains(":") ) return checkCommaAndDelimiter(str);
+        Matcher m = Pattern.compile("//(.)\n(.*)").matcher(str);
```
Effective Java 의 Item6 불필요한 객체 생성을 피하라 에 아래와 같이 나와있습니다.
Pattern은 입력받은 정규표현식으로 해당하는 유한 상태 머신(finite state machine)을 만들기 때문에 인스턴스 생성 비용이 높다.

따라서 문자열마다 패턴을 생성하기보다는 패턴자체를 상수로 선언하면 어떨까요?

codeleesh replied:
객체를 생성할 때 좀 더 주의깊게 해야겠습니다!

hhlyn0328 commented on src/main/java/util/StringAddCalculator.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    private static int checkInt(int number) {
+        if( number < 0 ) throw new RuntimeException("For input only positive integers");
```
보통 한 줄 의 코드도 중괄호 {} 로 감싸는 것을 컨벤션으로 많이 사용하는데요 
장점으로는 중괄호 {} 로 감싸지는 영역이 좀 더 명확히 구분되어 가독성이 좋고, 
구간 내에 코드가 추가 될 때, 괄호도 추가해주지 않아도 되는 것 정도? 일 것 같아요

하지만 가독성이라는 것이 주관적이라 반드시 중괄호 {} 의 사용이 정답이라고는 생각치 않습니다 
그럼에도 제가 만난 거의 모든 개발자들이 한 줄에도 중괄호 {} 를 사용하고 있었어서  
이렇게, 사용하시지 않는 분은 처음 뵙네요 :) 
개인적인 질문인데, 중괄호 {} 를 사용하지 않으신 이유를 문의드려도 될까요?
그냥 굳이? 써야하나 싶었다 라는 답변도 좋습니다 ! ㅎㅎ

codeleesh replied:
라인수를 줄여보려고 해보았는데, 저도 가독성은 좀 떨어진다고 느꼈습니다! 별다른 뜻은 없었습니다!

hhlyn0328 reviewed (changes requested):
문자열 덧셈 계산기 구현을 잘 해주셨네요 :+1: 

리뷰를 원하셨던 메인 로직에 대해 소소한 의견을 남겨보았습니다 
확인 부탁드려요 :)

hhlyn0328 reviewed (changes requested):
2단계 문자열 덧셈 계산기 구현을 잘 해주셨네요 👍 

리뷰를 원하셨던 메인로직에 대해 소소한 의견을 남겨보았습니다
확인 부탁드려요 :)

codeleesh commented:
> 2단계 문자열 덧셈 계산기 구현을 잘 해주셨네요 👍
> 
> 리뷰를 원하셨던 메인로직에 대해 소소한 의견을 남겨보았습니다 확인 부탁드려요 :)

리뷰 감사합니다!

codeleesh committed:
fix:
- 잘못된 형식의 문자열인 경우 예외 발생 처리
- Pattern을 이용하여 콤마 혹은 콜론[,:] 처리 및 상수 처리
- 콤마 외 상수 처리

codeleesh committed:
test: 음수 테스트의 메소드 수정

hhlyn0328 commented on src/main/java/util/StringAddCalculator.java:
```diff
@@ -5,49 +5,60 @@
+        if( result == ZERO ) {
+            throw new IllegalArgumentException(MESSAGE_INVALID_CHARACTER_FORMAT);
+        }
         return result;
```
```suggestion
        if( str == null || str.isEmpty() ) {
            return ZERO;
        }
        
        if( str.length() == 1 ) {
            result = checkOneNumber(str);
        }
        
        Matcher matcher = PATTERN_CUSTOM.matcher(str);
        if( matcher.find() ) {
            return checkCustomDelimiter(matcher);
        }
        
        return checkSeparatorAndDelimiter(str);
```

제가 말씀드렸던 부분은 이런 순서로 진행되면 될 것 같다는 부분이었습니다! 
`checkSeparatorAndDelimiter` 이게 기본 정책이어서, 
빈문자열 체크, 길이 하나 체크, 커스텀 체크 를 모두 통과하면 기본 정책에 따라 처리되어야 할테니까요 ! 

옳지 않은 형식의 문자열을 입력했다면 
아래의 기본정책 로직을 수행하면서 자연스럽게 예외가 발생할 것 같습니다 :) 
```
private static int checkSeparatorAndDelimiter(String str) {
         String[] token = PATTERN_DELIMITERS.split(str);
         return addStringArray(token);
 }
```

hhlyn0328 reviewed (approved):
리뷰사항 반영을 잘 해주셨네요 👍 👍 

부가적인 코멘트가 필요한 부분이 있는 것 같아 추가 코멘트 하나 남겼습니다! 
그럼, 다음 단계 진행을 위해 머지할게요 :)
